addons/test_mail/tests/test_ir_attachment.py

```diff
@@ -0,0 +1,76 @@
+            email="employee_second@example.com",
+            groups="base.group_user,base.group_partner_manager",
+        )
+
+        admin_attachments = self.env['ir.attachment'].with_user(user_admin).create([{
+            'type': 'binary',
+            'name': f'admin_attachment_{n}',
+            'datas': "ABC=",
+        } for n in range(2)])
+        employee_attachments = self.env['ir.attachment'].with_user(user_employee).create([{
+            'type': 'binary',
+            'name': f'employee_attachment_{n}',
+            'datas': "ABC=",
+        } for n in range(3)])
+
+        second_employee_attachments = self.env['ir.attachment'].with_user(user_second_employee).create({
+            'type': 'binary',
+            'name': 'second_employee_attachment',
+            'datas': "ABC=",
+        })
+
+        test_record.with_user(user_admin).message_post(body="Hi", attachment_ids=admin_attachments.ids)
+        test_record.with_user(user_employee).message_post(body="Hello", attachment_ids=employee_attachments.ids)
+        test_record.with_user(user_second_employee).sudo().message_post(body="Hello again", attachment_ids=employee_attachments[2].ids)
+        create_by_employee_send_by_second = employee_attachments[2]
+        test_record.with_user(user_second_employee).sudo().message_post(body="Hello again with own attachment", attachment_ids=second_employee_attachments.ids)
+
+        # forbidden
+        forbidden_list = [
+            (user_employee, admin_attachments[0]),
+            (user_employee, second_employee_attachments[0])
+        ]
+        for user, attachment in forbidden_list:
+            with self.subTest(user=user.name, attachment=attachment.name, method='write'):
+                with self.assertRaises(AccessError):
+                    attachment.with_user(user).write({'name': f'{user.name} {attachment.name[-1]}'})
+        for user, attachment in forbidden_list:
+            with self.subTest(user=user.name, attachment=attachment.name, method='unlink'):
+                with self.assertRaises(AccessError):
+                    attachment.with_user(user).unlink()
+
+        # allowed
+        allowed_list = [
+            (user_admin, admin_attachments[0], False),
+            (user_admin, employee_attachments[0], False),
+            (user_employee, admin_attachments[1], True),
+            (user_employee, employee_attachments[1], False),
+            (user_employee, create_by_employee_send_by_second, False),
+        ]
+        for user, attachment, sudo in allowed_list:
+            with self.subTest(user=user.name, attachment=attachment.name, sudo=sudo):
+                attachment.with_user(user).sudo(sudo).write({'name': f'{user.name} {attachment.name[-1]}'})
+        self.assertListEqual(admin_attachments.mapped('name'), ['Mitchell Admin 0', 'Ernest Employee 1'])
+        self.assertListEqual(employee_attachments.mapped('name'), ['Mitchell Admin 0', 'Ernest Employee 1', 'Ernest Employee 2'])
```

Comment:
Mmh, would make more sense to add the expected name inside the "allowed_list" and assert one by one, will make it easier to spot which one failed.

And it will make it part of the subTest, which is better IMO.

Reply:
Weird test indeed, looking back :raised_eyebrow: